ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java

```diff
@@ -96,74 +101,237 @@ public void close() {
         }
     }
 
-    private void init() {
-        class RefreshIngestionResourcesTask extends TimerTask {
-            @Override
-            public void run() {
-                try {
+    class RefreshIngestionResourcesTask extends TimerTask {
+        private boolean refreshedAtLeastOnce = false;
+
+        public boolean isRefreshedAtLeastOnce() {
+            return refreshedAtLeastOnce;
+        }
+
+        @Override
+        public void run() {
+            try {
+                MonitoredActivity.invoke((SupplierTwoExceptions<Void, IngestionClientException, IngestionServiceException>) () -> {
                     refreshIngestionResources();
-                    timer.schedule(new RefreshIngestionResourcesTask(), defaultRefreshTime);
-                } catch (Exception e) {
-                    log.error("Error in refreshIngestionResources. " + e.getMessage(), e);
-                    if (timer != null) {
-                        timer.schedule(new RefreshIngestionResourcesTask(), refreshTimeOnFailure);
+                    return null;
+                }, "ResourceManager.refreshIngestionResource");
+            } catch (Exception e) {
+                log.error("Error in refreshIngestionResources. " + e.getMessage(), e);
+                scheduleRefreshIngestionResourcesTask(refreshTimeOnFailure);
+            }
+        }
+
+        private void refreshIngestionResources() throws IngestionClientException, IngestionServiceException {
```

Comment:
All these methods were on the `ResourceManager` class, though now `RefreshIngestionResourcesTask` is effectively their wrapper with retry logic, so they belong there.